flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/parser/SqlClientSyntaxHighlighter.java

```diff
@@ -0,0 +1,246 @@
+ */
+
+package org.apache.flink.table.client.cli.parser;
+
+import org.apache.flink.table.api.SqlDialect;
+import org.apache.flink.table.api.config.TableConfigOptions;
+import org.apache.flink.table.client.gateway.Executor;
+
+import org.jline.reader.LineReader;
+import org.jline.reader.impl.DefaultHighlighter;
+import org.jline.utils.AttributedString;
+import org.jline.utils.AttributedStringBuilder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Properties;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.client.cli.CliClient.COLOR_SCHEMA_VAR;
+
+/** Sql Client syntax highlighter. */
+public class SqlClientSyntaxHighlighter extends DefaultHighlighter {
+    private static final Set<String> FLINK_KEYWORD_SET;
+    private static final Set<String> HIVE_KEYWORD_SET;
+
+    static {
+        try (InputStream is =
+                SqlClientSyntaxHighlighter.class.getResourceAsStream("/keywords.properties")) {
+            Properties props = new Properties();
+            props.load(is);
+            FLINK_KEYWORD_SET =
+                    Collections.unmodifiableSet(
+                            Arrays.stream(props.get("default").toString().split(";"))
+                                    .collect(Collectors.toSet()));
+            HIVE_KEYWORD_SET =
+                    Collections.unmodifiableSet(
+                            Arrays.stream(props.get("hive").toString().split(";"))
+                                    .collect(Collectors.toSet()));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private final Executor executor;
+
+    public SqlClientSyntaxHighlighter(Executor executor) {
+        this.executor = executor;
+    }
+
+    @Override
+    public AttributedString highlight(LineReader reader, String buffer) {
+
+        final Object colorSchemeOrdinal = reader.getVariable(COLOR_SCHEMA_VAR);
+        SyntaxHighlightStyle.BuiltInStyle style =
+                SyntaxHighlightStyle.BuiltInStyle.fromOrdinal(
+                        colorSchemeOrdinal == null ? 0 : (Integer) colorSchemeOrdinal);
+        if (style == SyntaxHighlightStyle.BuiltInStyle.DEFAULT) {
+            return super.highlight(reader, buffer);
+        }
+        final SqlDialect dialect =
+                executor.getSessionConfig()
+                                .get(TableConfigOptions.TABLE_SQL_DIALECT)
+                                .equalsIgnoreCase(SqlDialect.HIVE.toString())
+                        ? SqlDialect.HIVE
+                        : SqlDialect.DEFAULT;
+        return getHighlightedOutput(buffer, style.getHighlightStyle(), dialect);
+    }
+
+    static AttributedString getHighlightedOutput(
```

Comment:
Can we do like this?
```
switch (currentChar) {
                case '\'':
                    if (state == State.SINGLE_QUOTE) {
                        state = State.NORMAL;
                    } else if (state == State.NORMAL) {
                        state = State.SINGLE_QUOTE;
                    }
                    break;
                case '"':
                    if (state == State.DOUBLE_QUOTE) {
                        state = State.NORMAL;
                    } else if (state == State.NORMAL) {
                        state = State.DOUBLE_QUOTE;
                    }
                    break;
                case '`':
                    if (state == State.BACK_QUOTE) {
                        state = State.NORMAL;
                    } else if (state == State.NORMAL) {
                        state = State.BACK_QUOTE;
                    }
                    break;
                case '-':
                    if (lastChar == '-' && state == State.NORMAL) {
                        state = State.SINGLE_COMMENT;
                    }
                    break;
                case '\n':
                    if (state == State.SINGLE_COMMENT) {
                        state = State.NORMAL;
                    }
                    currentPaddingLineBuilder.setLength(0);
                    currentPaddingSqlBuilder.append("\n");
                    break;
                case '*':
                    if (lastChar == '/' && state == State.NORMAL) {
                        state = State.MULTI_LINE_COMMENT;
                    }
                    break;
                case '/':
                    if (lastChar == '*' && state == State.MULTI_LINE_COMMENT) {
                        state = State.NORMAL;
                    }
                    break;
```

After state is determined, then we apply the style on to the words.

Reply:
I extracted state determination in a separate method and then use it to apply it on the words.

Reply:
In general it shows the behavior howbeit there are some details like splitting words.
In fact if there is a `DEFAULT` state and current symbol is not a part of word, e.g. whitespace or some other character then it's better to append it directly to output. This will allow to not care about any possible whitespace or non word related prefixes while checking against keywords